pyro/infer/autoguide/guides.py

```diff
@@ -1190,3 +1188,280 @@ def forward(self, *args, **kwargs):
+                    _deep_setattr(self.scale_trils, name,
+                                  PyroParam(init_scale_tril, self.scale_tril_constraint))
+
+            # Initialize dependencies on upstream variables.
+            num_pending[name] = 0
+            deps = PyroModule()
+            _deep_setattr(self.deps, name, deps)
+            for upstream, dep in self.dependencies.get(name, {}).items():
+                assert upstream in self.prototype_trace.nodes
+                children[upstream].append(name)
+                num_pending[name] += 1
+                if isinstance(dep, str) and dep == "linear":
+                    dep = torch.nn.Linear(numel[upstream], numel[name], bias=False)
+                    dep.weight.data.zero_()
+                elif not callable(dep):
+                    raise ValueError(
+                        f"Expected either the string 'linear' or a callable, but got {dep}"
+                    )
+                _deep_setattr(deps, upstream, dep)
+
+        # Topologically sort sites.
+        self._sorted_sites = []
+        while num_pending:
+            name, count = min(num_pending.items(), key=lambda kv: (kv[1], kv[0]))
+            assert count == 0, f"cyclic dependency: {name}"
+            del num_pending[name]
+            for child in children[name]:
+                num_pending[child] -= 1
+            self._sorted_sites.append((name, self.prototype_trace.nodes[name]))
+
+    @poutine.infer_config(config_fn=_config_auxiliary)
+    def get_deltas(self, save_params=None):
+        deltas = {}
+        aux_values = {}
+        for name, site in self._sorted_sites:
+            if save_params is not None and name not in save_params:
+                continue
+
+            # Sample zero-mean blockwise independent Delta/Normal/MVN.
+            loc = _deep_getattr(self.locs, name)
+            zero = torch.zeros_like(loc)
+            conditional = self.conditionals[name]
+            if callable(conditional):
+                aux_value = _deep_getattr(self.conds, name)()
+            elif conditional == "delta":
+                aux_value = zero
+            elif conditional == "normal":
+                scale = _deep_getattr(self.scales, name)
+                aux_value = pyro.sample(
+                    name + "_aux",
+                    dist.Normal(zero, scale).to_event(1),
+                    infer={"is_auxiliary": True},
+                )
+            elif conditional == "mvn":
+                # This overparametrizes by learning (scale,scale_tril),
+                # enabling faster learning of the more-global scale parameter.
+                scale = _deep_getattr(self.scales, name)
+                scale_tril = _deep_getattr(self.scale_trils, name)
+                scale_tril = scale[..., None] * scale_tril
+                aux_value = pyro.sample(
+                    name + "_aux",
+                    dist.MultivariateNormal(zero, scale_tril=scale_tril),
```

Comment:
If we factor this out to `scale_tril @ Normal(0, 1)`, I guess HMC will be a bit happier.

Reply:
Great point, I guess that is equivalent to reparametrizing.  I've also changed `Normal(0,scale)` to `Normal(0,1) * scale` in the "normal" case.

Reply:
I think you will need to add [logdet](https://github.com/pyro-ppl/pyro/blob/6ec3be8186fb669cdb2dc797920444ba0e5904ca/pyro/distributions/transforms/lower_cholesky_affine.py#L66) of those affine transforms. How about using `dist.TransformedDistribution(dist.Normal(...), LowerCholeskyAffine(...))` so that we can use `TransformReparam` in the reparam?

Reply:
Thanks, I've added the logdet terms by hand here since it's simpler.  Does it look right now?

Reply:
Yes, it looks correct to me.

Reply:
I believe two versions are equivalent... Not sure what's going on. Let me play with some tests to see if elbo is the same for the two versions.

Reply:
Thanks I'll do the same, at least to create a unit test I can run locally (not on some huge model on a GPU cloud machine)

Reply:
I think I find the issue. Here `log_density` is calculated as the sum over all dimensions of the site. However, the `ldj` term, which is used to calculate the logdet of unconstrained->constrained values, maintains the batch dimension. So sum of them will give wrong result if this site is under some plate. I guess we should use `pyro.factor` for those `log_density` terms. What do you think?

Reply:
@fehiepsi thanks, yes I now see the error.  I'll think about this and submit a fix ASAP.